### DISCUSSION TITLE
Prism for WinUI 3

### DISCUSSION THREAD

**Discussion by odp-remec** [PrismLibrary/Prism#2675] [General]
Hi,

At our company we use Prism in several products built with WPF. Prism works well for us and fits our needs.
Our newest product is built with WinUI 3, in which we like to use Prism as well.

Using the Uno implementation of Prism for WinUI is not suitable for us, because we don’t want to have a dependency on the Uno Platform in a plain WinUI project. So we took the WPF implementation and extended this version to support WinUI. The WPF implementation already contains preprocessor directives with the symbol `HAS_WINUI`, which we could use and extend.

Before we submit a Pull-Request we would like to discuss with you if you are generally interested in a WinUI version and if our solution fits the Prism requirements. We are currently using this version in a beta version of our software, so far it works well for us. You can have a look at our Prism WinUI solution at the following link: https://github.com/bennoremec/PrismWinUi

Please be aware that we left out some features that are currently not important for us. This includes the following features:

- No implementation for DryIoc. We currently use Unity in all our Prism projects. Due to the deprecated state of Unity.Container we may change this in the future and switch to DryIoc.
- DialogService or rather the folder Services/Dialogs is not implemented for WinUI because we do not need Prism Dialogs.
- No implementation for PrismApplication and PrismApplicationBase. PrismBootstrapper should be used.
- DirectoryModuleCatalog implemented for NetCore but not for Net45.
- XamlModuleCatalog is not implemented.
- Tests are missing
  - Prism.Core.Tests
  - Prism.WinUi.Tests
  - Prism.Unity.WinUi.Tests

In case of a Pull-Request, we would of course add certain parts, especially the tests.
Please let us know what you think about this.

**Comment by LeonSpors** [comment]
Why do you have made the shell unnecessary by adding `GetMainWindow()`. 
It's kinda weird because you have to define a **DependencyObject** as shell but it is never used. 
The window is used in `OnInitialized()`.

In my test project the regions are added to the **DependencyObject** (shell) but not to the window target. This means that we are not able to navigate.

I hope you can clear this out and I appreciate the work you have done for all of us.

**Reply by odp-remec** [reply]
Thank you for your reply @LeonSpors

We use the `PrismBootstrapperBase` class from the WPF implementation. In WPF, shell is of type window and window is a DependencyObject. This works for WPF but not for WinUI. In WinUI, window is not a DependecyObject. 
As a workaround we created `MainWindow.xaml` of type `Window` and added the shell as content. To activate the window in `OnInitialized()` we need `GetMainWindow()` to get access to the window instance.
I agree with you that this is not a good solution.

I have pushed an alternative version on our fork: https://github.com/bennoremec/PrismWinUi/commit/2caded7bcd27be99f2a1ffd01c6982ff15260156
You can now use PrismApplication instead of PrismBootstrapper. This should work the same way as with WPF. 
See the example here: https://prismlibrary.com/docs/wpf/getting-started.html#override-the-existing-application-object
In contrast to WPF, Shell is of type UIElement. This now works similar to the Uno Version of Prism.
You must override `CreateShell()` which has as return type `UIElement` and  your Shell must be a UserControl.

Please let me know if this works for you.

**Comment by soroshsabz** [comment]
ITNOA

@dansiegel Any update for this? Did you any plan to support WinUI3 and Win App SDK natively?

**Reply by brianlagunas** [reply]
There is no update. There are no immediate plans to create a version of Prism for WinUI 3.

**Reply by soroshsabz** [reply]
Thanks for responding, Are you accept PR for this?

**Reply by brianlagunas** [reply]
No, we are not at this time
